core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,69 +1327,337 @@ class ReassignedPartitionsIsrChangeListener(protected val controller: KafkaContr
-
-  def doHandleChildChange(parentPath: String, currentChildren: Seq[String]): Unit = {
-    inLock(controller.controllerContext.controllerLock) {
-      debug("ISR change notification listener fired")
+  case class IsrChangeNotification(sequenceNumbers: Seq[String]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
       try {
-        val topicAndPartitions = currentChildren.flatMap(getTopicAndPartition).toSet
+        val topicAndPartitions = sequenceNumbers.flatMap(getTopicAndPartition).toSet
         if (topicAndPartitions.nonEmpty) {
-          controller.updateLeaderAndIsrCache(topicAndPartitions)
+          updateLeaderAndIsrCache(topicAndPartitions)
           processUpdateNotifications(topicAndPartitions)
         }
       } finally {
-        // delete processed children
-        currentChildren.map(x => controller.controllerContext.zkUtils.deletePath(
-          ZkUtils.IsrChangeNotificationPath + "/" + x))
+        // delete the notifications
+        sequenceNumbers.map(x => controllerContext.zkUtils.deletePath(ZkUtils.IsrChangeNotificationPath + "/" + x))
+      }
+    }
+
+    private def processUpdateNotifications(topicAndPartitions: immutable.Set[TopicAndPartition]) {
+      val liveBrokers: Seq[Int] = controllerContext.liveOrShuttingDownBrokerIds.toSeq
+      debug("Sending MetadataRequest to Brokers:" + liveBrokers + " for TopicAndPartitions:" + topicAndPartitions)
+      sendUpdateMetadataRequest(liveBrokers, topicAndPartitions)
+    }
+
+    private def getTopicAndPartition(child: String): Set[TopicAndPartition] = {
+      val changeZnode: String = ZkUtils.IsrChangeNotificationPath + "/" + child
+      val (jsonOpt, _) = controllerContext.zkUtils.readDataMaybeNull(changeZnode)
+      if (jsonOpt.isDefined) {
+        val json = Json.parseFull(jsonOpt.get)
+
+        json match {
+          case Some(m) =>
+            val topicAndPartitions: mutable.Set[TopicAndPartition] = new mutable.HashSet[TopicAndPartition]()
+            val isrChanges = m.asInstanceOf[Map[String, Any]]
+            val topicAndPartitionList = isrChanges("partitions").asInstanceOf[List[Any]]
+            topicAndPartitionList.foreach {
+              case tp =>
+                val topicAndPartition = tp.asInstanceOf[Map[String, Any]]
+                val topic = topicAndPartition("topic").asInstanceOf[String]
+                val partition = topicAndPartition("partition").asInstanceOf[Int]
+                topicAndPartitions += TopicAndPartition(topic, partition)
+            }
+            topicAndPartitions
+          case None =>
+            error("Invalid topic and partition JSON: " + jsonOpt.get + " in ZK: " + changeZnode)
+            Set.empty
+        }
+      } else {
+        Set.empty
+      }
+    }
+
+  }
+
+  case class PreferredReplicaLeaderElection(partitionsForPreferredReplicaElection: Set[TopicAndPartition]) extends ControllerEvent {
```

Comment:
What's the thinking around making these inner classes? It seems like it was better when it was clear when controller methods were being invoked?